src/opt/ReplaceHalfMaskedMemOps.cpp

```diff
@@ -0,0 +1,216 @@
+    auto *firstVecType = llvm::cast<llvm::VectorType>(firstVector->getType());
+    auto *secondVecType = llvm::cast<llvm::VectorType>(secondVector->getType());
+    unsigned N = firstVecType->getElementCount().getKnownMinValue();
+
+    assert(firstVecType == secondVecType);
+
+    // Create a vector of indices to merge two vectors
+    std::vector<llvm::Constant *> indices;
+    for (unsigned i = 0; i < N * 2; ++i) {
+        indices.push_back(llvm::ConstantInt::get(B.getInt32Ty(), i));
+    }
+    llvm::ArrayRef<llvm::Constant *> indicesRef(indices);
+    llvm::Constant *mask = llvm::ConstantVector::get(indicesRef);
+
+    // Use shufflevector instruction to create the new vector
+    return B.CreateShuffleVector(firstVector, secondVector, mask, name);
+}
+
+llvm::Value *lShrinkVector(llvm::IRBuilder<> &B, llvm::Value *originalVector) {
+    auto *vecType = llvm::cast<llvm::VectorType>(originalVector->getType());
+    unsigned N = vecType->getElementCount().getKnownMinValue();
+
+    // Create a vector of indices to select the first half of the elements
+    std::vector<llvm::Constant *> indices;
+    for (unsigned i = 0; i < N / 2; ++i) {
+        indices.push_back(llvm::ConstantInt::get(B.getInt32Ty(), i));
+    }
+    llvm::ArrayRef<llvm::Constant *> indicesRef(indices);
+    llvm::Constant *mask = llvm::ConstantVector::get(indicesRef);
+
+    // Use shufflevector instruction to create the new vector
+    llvm::Twine name = originalVector->getName() + ".half";
+    return B.CreateShuffleVector(originalVector, llvm::UndefValue::get(vecType), mask, name);
+}
+
+llvm::Value *lBitcastPointerType(llvm::IRBuilder<> &B, llvm::Value *ptr, llvm::Value *value) {
+    auto *vecType = llvm::cast<llvm::VectorType>(value->getType());
+    auto *newPtrType = llvm::PointerType::get(vecType, 0 /* TODO! */);
+    // TODO! opaque pointer is no-op here, any special handling?
+    return B.CreateBitCast(ptr, newPtrType);
+}
+
+llvm::Constant *lShrinkConstVec(llvm::LLVMContext &context, llvm::Value *originalValue) {
+    llvm::Constant *originalMask = llvm::cast<llvm::Constant>(originalValue);
+    auto *vecType = llvm::cast<llvm::VectorType>(originalMask->getType());
+    unsigned N = vecType->getElementCount().getKnownMinValue();
+    unsigned newSize = N / 2;
+
+    // Extract the second half of the elements
+    std::vector<llvm::Constant *> newValues;
+    for (unsigned i = newSize; i < N; ++i) {
+        auto *element = originalMask->getAggregateElement(i);
+        if (!element) {
+            // Handle error in extracting element
+            return nullptr;
+        }
+        newValues.push_back(element);
+    }
+
+    // Create a new constant vector with the first half of the elements
+    llvm::ArrayRef<llvm::Constant *> newValuesRef(newValues);
+    return llvm::ConstantVector::get(newValuesRef);
+}
+
+// This function replaces, e.g.,
+//
+// %ptr = bitcast %v8_uniform_FVector4f* %Result.i to <8 x float>*
```

Comment:
What would be code generation if the original `N = 4` and we cut it to `load <2 x float>`? We need to ensure it wouldn't be worse than just `masked.load/store.v4f32`.